torch/csrc/jit/codegen/cuda/lower_double_buffer.h

```diff
@@ -77,6 +77,85 @@
 // - Double the allocation size
 // - Omit the RAW sync in the Main and Epilogue loops
 
+// [Cicular buffer] An generalization of double buffering.
+// On sm80+ hardware there is asynchronous copy infrastructure that
+//  motivates a circular buffering generalization of double buffering.
+// Almost all analyses previously done for double buffering are exactly
+//  the same with circular buffering, except for the introduction of
+//  new concept: `stage depth`.
+//
+// The `stage depth` is defined as the multiplier of extra buffering
+//  space used. In the case of double buffering, the stage depth would
+//  be 2.
+//
+// A circular buffered loop structure would look like follows, which
+//  exactly parallels the case of double buffered loop structure, since
+//  it is a exact generalization to the same purpose.
+//
+// Here S is the original allocation size as above,
+//  D is the stage depth. With D=2, the below loop structure becomes
+//  exactly the same as the case in double buffering.
+//
+// allocate X[S*D] // allocation
+// for i in 0..D-1: // prolog
+//   for j in ...
+//     if pred:
+//       x[i*S+j] = y[i, j];
+//
+// for i in 0..N: // main loop
+//   for j in ...
+//     if pred:
+//       x[((i+D-1)%D)*S+j] = y[i+D-1, j];
+//   for j in ...
+//     .. = x[(i%D)*S+j]
+//
+// (Epilog omitted since this only makes sense in using
+// cp.async, where producer will be in global mem and consumer will
+// be in shared mem).
```

Comment:
I'm not sure I follow this, isn't it just not necessary based on how you set the loops/predicates up?

Reply:
The current double buffer/circular buffer infra would generate an epilog if the double buffered tensor is loading from shared mem. This comment just saying cp.async would only read from gmem so in this case there'd not be an epilog. 

It could change over time.